domain/machine/state/machine_cloud_instance.go

```diff
@@ -192,13 +193,110 @@ WHERE m.name = $machineName.name;
 	return instanceId, nil
 }
 
-// InstanceStatus returns the cloud specific instance status for this
+// GetInstanceStatus returns the cloud specific instance status for the given
 // machine.
-// If the machine is not provisioned, it returns a NotProvisionedError.
-func (st *State) InstanceStatus(ctx context.Context, machineName machine.Name) (string, error) {
-	// TODO(cderici): Implementation for this is deferred until the design for
-	// the domain entity statuses on dqlite is finalized.
-	return "running", nil
+// It returns a StatusNotSet if the instance status is not set.
+// Idempotent.
+func (st *State) GetInstanceStatus(ctx context.Context, mName machine.Name) (status.Status, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	machineStatus := machineInstanceStatus{Name: mName}
+	statusQuery := `
+SELECT mcis.status as &machineInstanceStatus.status
+FROM machine as m
+	JOIN machine_cloud_instance_status as mcis ON m.uuid = mcis.machine_uuid
+WHERE m.name = $machineInstanceStatus.name;
+`
+
+	statusQueryStmt, err := st.Prepare(statusQuery, machineStatus)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, statusQueryStmt, machineStatus).Get(&machineStatus)
+		if err != nil {
+			return errors.Annotatef(err, "querying cloud instance status for machine %q", mName)
+		}
+		return nil
+	})
+
+	if err != nil {
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return "", errors.Annotatef(machineerrors.StatusNotSet, "machine: %q", mName)
+		}
+		return "", errors.Trace(err)
+	}
+	internalStatus := machineStatus.Status
+	// Convert the internal status id from the (instance_status_values table)
+	// into the core status.Status type.
+	var instanceStatus status.Status
+	switch internalStatus {
+	case 0:
+		instanceStatus = status.Empty
+	case 1:
+		instanceStatus = status.Allocating
+	case 2:
+		instanceStatus = status.Running
+	case 3:
+		instanceStatus = status.ProvisioningError
+	}
+	return instanceStatus, nil
+}
+
+// SetInstanceStatus sets the cloud specific instance status for this
+// machine.
+func (st *State) SetInstanceStatus(ctx context.Context, mName machine.Name, instanceStatus status.Status) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	var iStatus int
+	switch instanceStatus {
+	case status.Empty:
+		iStatus = 0
+	case status.Allocating:
+		iStatus = 1
+	case status.Running:
+		iStatus = 2
+	case status.ProvisioningError:
+		iStatus = 3
+	}
```

Comment:
There should be a test that checks these values against the database values and probably error out if the status values don't match.